### DISCUSSION TITLE
STRICT tables and type conversion

### DISCUSSION THREAD

**Discussion by kuhnroyal** [simolus3/drift#1651] [Ideas]
I started using the new `STRICT` table feature from sqlite3.37 and already discovered some potential errors where some columns with unknown data types made it into the database. So this is really good, especially when writing migrations with SQL.

We tend to copy/paste the SQL from drift files to migrations and adapt it, here we forget converting from `BOOLEAN` to `INTEGER` etc. but now this is prevented.

But when making tables `STRICT` in drift files, there are a lot of warnings about unknown types which get converted by Drift into the correct types - `BOOLEAN`, `DATETIME`, `DECIMAL` etc.

So maybe it would be helpful to have something similar to the `MAPPED BY` syntax in order to prevent these warning:

```sql
-- Generates warning
foo BOOLEAN DEFAULT FALSE,
-- Would not generate warning and can be copied to migration without changing the type
foo INTEGER DEFAULT 0 MAPPED TO BOOLEAN,
```

This would prevent the STRICT warning and can easier be copied to an SQL migration.

What do you think?

**Comment by simolus3** [comment]
I think the first `foo` should not generate a warning either. It does at the moment, but that's the implementation being overly strict. At the moment, we allow any of the following in `sqlparser`

- `INT`, `INTEGER`, `REAL`, `TEXT`, `BLOB` and `ANY` all the time (following [sqlite docs](https://www.sqlite.org/stricttables.html#strict_tables), but support for `ANY` is limited at the moment)
- additionally, `BOOL` and `DATE` if the `CREATE TABLE` statement being analyzed comes from a drift file

So using say `DECIMAL` instead of `REAL` should be runtime error (if I read sqlite3 docs correctly) and is flagged by the static analyzer too. For tables defined in drift files, I think we can be more lenient and allow anything _containing_ `BOOL` or `DATE`. It will be turned into an int by drift either way.

Do you copy tables from drift files into raw SQL strings that you run in a migration? I don't really see the advantage of the `MAPPED TO BOOLEAN` syntax except that it would be easier to see and remove in that case.

**Reply by kuhnroyal** [reply]
> I think the first foo should not generate a warning either.

I agree, types that Drift correctly converts should not generate a warning.

> Do you copy tables from drift files into raw SQL strings that you run in a migration? I don't really see the advantage of the MAPPED TO BOOLEAN syntax except that it would be easier to see and remove in that case.

Yes that is exactly what I do when I want to change a column or remove it. The new syntax is probably not needed if we can get rid of the red warnings in the generator log :)
The copy/paste mistakes are probably not an issue when using `STRICT` in the migration.

**Reply by simolus3** [reply]
Thanks for the idea, I've improved the handling for drift files in aa38ab42f2cf7cc8279dc1809a5525d3b7bdce2a.

**Reply by kuhnroyal** [reply]
Awesome, i'll test this later or tomorrow!

**Reply by kuhnroyal** [reply]
This fails for `REAL|DECIMAL` types, there is no handling in `_handleColumnTypeWithoutDefault`.
Tried to adapt this but I am also not sure about the affinity `from_create_table_test`:
```dart
  'BOOLEAN': BasicType.real, // int?
  'DATE': BasicType.real, // int?
  'DATETIME': BasicType.real, // int?
```

**Reply by simolus3** [reply]
Good catch, thanks!

Outside of drift files, I think  `DECIMAL` should give an analysis error, it fails for me in sqlite3 (using `3.37.2`):

```
sqlite> CREATE TABLE tmp (x DECIMAL) STRICT;
Error: in prepare, unknown datatype for tmp.x: "DECIMAL" (1)
```

The truth is that drift's generator will _always_  pick some column type (falling back to `REAL` if needed). So the right behavior is probably just to completely omit the warning for drift files.

> Tried to adapt this but I am also not sure about the affinity `from_create_table_test`:

These refer to how the type names would parse outside of drift files, and are pretty much taken from the [sqlite3 documentation](https://www.sqlite.org/datatype3.html#affinity_name_examples). The `supports booleans when moor extensions are enabled` test is handling the other scenario.

**Reply by kuhnroyal** [reply]
> Outside of drift files, I think DECIMAL should give an analysis error

Yes, think so. I have been using this in Drift files so far but with STRICT, I will migrate to using `REAL` directly.

> The truth is that drift's generator will always pick some column type (falling back to REAL if needed).

Yea, I saw the fallback. Maybe this should move into the `_handleColumnTypeWithoutDefault` function.

**Reply by kuhnroyal** [reply]
Ah nvm, you did that already :) I'll test.

**Reply by kuhnroyal** [reply]
Works! Thanks! Can you release a patch version of the sqlparser?

**Reply by simolus3** [reply]
Thanks for all your help testing! Just published version `0.19.2` of `sqlparser`.
